apps/workshop/projects/AllbridgeCore/modules/AllbridgePool.ts

```diff
@@ -0,0 +1,95 @@
+import type {
+  ModuleDefinitionInterface,
+  Address,
+  Context,
+  Pool,
+  FetchPoolsContext,
+  FetchUserPositionsContext,
+  UserPosition
+} from '@defiyield/sandbox';
+import {getTokenInfo, getTokenInfos} from '../helpers/token-info';
+import poolAbi from '../abis/pool-abi.json'
+import {ChainSymbol, chainSymbolToSupportedChain} from '../helpers/chain';
+import {BigNumber} from 'ethers';
+
+const SYSTEM_PRECISION = 1e3;
+export function getAllbridgeModule(chainSymbol: ChainSymbol): ModuleDefinitionInterface {
+  return {
+    name: `Allbridge Core ${chainSymbol}`,
+    chain: chainSymbolToSupportedChain(chainSymbol),
+    type: 'pools',
+
+    /**
+     * Fetches the addresses of all involved tokens (supplied, rewards, borrowed, etc)
+     *
+     * @param ctx
+     * @returns Address[]
+     */
+    async preloadTokens(ctx: Context): Promise<Address[]> {
+      const tokenInfo = await getTokenInfos(ctx, chainSymbol)
+      return tokenInfo.map(info => info.tokenAddress);
+    },
+
+    /**
+     * Returns full pool list
+     *
+     * @param ctx
+     * @returns Pool[]
+     */
+    async fetchPools(ctx: FetchPoolsContext): Promise<Pool[]> {
+      const result: Pool[] = [];
+      for (const token of ctx.tokens) {
+        const tokenInfo = await getTokenInfo(ctx, chainSymbol, token.address);
+        if (!tokenInfo) {
+          continue
+        }
+
+        const poolContract = new ctx.ethcall.Contract(tokenInfo.poolAddress, poolAbi);
+        const [tokenBalance] = await ctx.ethcallProvider.all<typeof BigNumber>([poolContract.tokenBalance()])
```

Comment:
could you combine these so that theres a single multicall outside of the loop? something along the lines of this maybe
```ts
// filter out tokens to make looping simpler later
const tokens = []
const tokenInfos = await getTokenInfos(ctx, chainSymbol)
for (const token of ctx.tokens) {
  const tokenInfo = tokenInfos.find(t => t.tokenAddress.toLowerCase() === token.address.toLowerCase())
  if (!tokenInfo) {
    continue
  }

  tokens.push({ token: token, info: tokenInfo })
}

// queue up all balance requests
const calls = []
for (const token of tokens) {      
  const poolContract = new ctx.ethcall.Contract(token.info.poolAddress, poolAbi);
  calls.push(poolContract.tokenBalance())
}

// get balance for every token in a single rpc call
const tokenBalances = await ctx.ethcallProvider.all<BigNumber>(calls)

// carry on and push to results like before
for (const token of tokens) {
  const tvl = (+tokenBalances.shift()!) / SYSTEM_PRECISION;
  const pool: Pool = {
    id: token.info.poolAddress,
```